prost-build/src/lib.rs

```diff
@@ -575,9 +608,49 @@ impl Config {
             file.flush()?;
         }
 
+        if let Some(ref include_file) = self.include_file {
+            trace!("Writing include file: {:?}", target.join(include_file));
+            let mut file = fs::File::create(target.join(include_file))?;
+            self.write_includes(modules.keys().collect(), &mut file, 0, if target_is_env { None } else { Some(&target) })?;
+            file.flush()?;
+        }
+
         Ok(())
     }
 
+    fn write_includes(&self, mut entries: Vec<&Module>, outfile: &mut fs::File, depth: usize, basepath: Option<&PathBuf>) -> Result<usize> {
+        let mut written = 0;
+        while entries.len() > 0 {
+            let modident = &entries[0][depth];
+            let matching: Vec<&Module> = entries.iter().filter(|&v| &v[depth] == modident).map(|v| *v).collect();
+            {
+                // Will NLL sort this mess out?
+                let _temp = entries.drain(..).filter(|&v| &v[depth] != modident).collect();
+                entries = _temp;
+            }
+            self.write_line(outfile, depth, &format!("pub mod {} {{", modident))?;
+            let subwritten = self.write_includes(matching.iter().filter(|v| v.len() > depth + 1).map(|v| *v).collect(),
+                                                 outfile, depth + 1, basepath)?;
+            written += subwritten;
+            if subwritten != matching.len() {
+                let modname = matching[0][..=depth].join(".");
+                if let Some(buf) = basepath {
+                    self.write_line(outfile, depth + 1, &format!("include!(\"{:?}/{}.rs\");", buf, modname))?;
+                } else {
+                    self.write_line(outfile, depth + 1, &format!("include!(concat!(env!(\"OUT_DIR\"), \"/{}.rs\"));", modname))?;
+                }
+                written += 1;
+            }
+
+            self.write_line(outfile, depth, "}")?;
+        }
+        Ok(written)
+    }
+
+    fn write_line(&self, outfile: &mut fs::File, depth: usize, line: &str) -> Result<()> {
+        outfile.write_all(format!("{}{}\n", ("    ").to_owned().repeat(depth), line).as_bytes())
+    }
```

Comment:
Not sure about @danburkert 's preference, but IMO since this is generated (and probably temporary) code, it is not worth the effort to try and format it like this. Instead just make something that parses. If the user wants it formatted, they can use rustfmt.

Reply:
I believe (and this was a while ago :-)) that I did this because the main code output by the generator is reasonably formatted as well.